proof/refine/ARM/Ipc_R.thy

```diff
@@ -2053,23 +2060,36 @@ lemma bind_sc_reply_invs[wp]:
   apply (erule (1) valid_objs_sc_replies_distinct)
   done
 
+crunches bind_sc_reply
+  for pspace_distinct[wp]: pspace_distinct
+  and pspace_aligned[wp]: pspace_aligned
+
+lemma bind_sc_reply_valid_objs[wp]:
+  "\<lbrace>valid_objs and reply_at reply_ptr and sc_at sc_ptr and
+    sc_replies_sc_at (\<lambda>a. reply_ptr \<notin> set a) sc_ptr\<rbrace>
+   bind_sc_reply sc_ptr reply_ptr
+   \<lbrace>\<lambda>_. valid_objs\<rbrace>"
+  unfolding bind_sc_reply_def
+  apply (wpsimp wp: hoare_list_all_lift)
+  apply (drule (1) valid_objs_ko_at)
+  by (clarsimp simp: valid_obj_def valid_sched_context_def sc_at_pred_n_def obj_at_def)
+
 lemma replyPush_corres:
   "can_donate = can_donate' \<Longrightarrow>
-   corres dc (invs and tcb_at caller and tcb_at callee and reply_at reply_ptr
-              and ex_nonz_cap_to reply_ptr
-              and st_tcb_at active caller
-              and reply_sc_reply_at (\<lambda>scp. scp = None) reply_ptr
-              and reply_tcb_reply_at (\<lambda>tptr. tptr = None) reply_ptr
-              and weak_valid_sched_action and scheduler_act_not caller)
-             (valid_release_queue_iff and valid_objs' and valid_queues and valid_queues')
+   corres dc (valid_replies and pspace_aligned and pspace_distinct and valid_objs and
+              (\<lambda>_. caller \<noteq> idle_thread_ptr) and tcb_at caller and tcb_at callee and
```

Comment:
Since we're discussing style guide stuff, is there a preference for having connectives at the end of the line versus the beginning of the next line? I prefer having them at the beginning of the line so I don't have to hunt back to the end of the previous line to double check what the connective is

Reply:
I agree.

Reply:
This is one which might be hard to get complete consensus on. Personally, I agree with both of you with having connectives like `and` and `|` at the start but at the same time still prefer things like `,` and `;` at the end.

Reply:
One could say "logical connectives should be at the start of new lines with the exception of puncuation (e.g. `,` and `;`"

Reply:
That would be my style but I know that some others like all connectives to go at the start, including `,` and `;`. As is usually the case, we're not going to make everyone happy so at some point we'll just need to pick an approach and stick with it until we all adjust.